homeassistant/helpers/entity_component.py

```diff
@@ -131,13 +115,91 @@ def _setup_platform(self, platform_type, platform_config,
         if platform is None:
             return
 
+        if self.group is None and self.group_name is not None:
+            self.group = group.Group(self.hass, self.group_name,
+                                     user_defined=False)
+
+        platform_name = '{}.{}'.format(self.domain, platform_type)
+        if platform_name in self.entity_platform_components:
+            platform_component = self.entity_platform_components[platform_name]
+        else:
+            scan_interval = platform_config.get(CONF_SCAN_INTERVAL,
+                                                self.scan_interval)
+            platform_component = EntityPlatformComponent(
+                self.logger, self.hass, self.entity_id_format,
+                self.group, scan_interval, platform_name)
+            self.entity_platform_components[platform_name] = platform_component
+
         try:
             platform.setup_platform(
-                self.hass, platform_config, self.add_entities, discovery_info)
+                self.hass, platform_config,
+                platform_component.add_entities, discovery_info)
         except Exception:  # pylint: disable=broad-except
             self.logger.exception(
                 'Error while setting up platform %s', platform_type)
             return
 
-        platform_name = '{}.{}'.format(self.domain, platform_type)
         self.hass.config.components.append(platform_name)
+
+
+class EntityPlatformComponent(object):
+    # pylint: disable=too-many-instance-attributes
+    # pylint: disable=too-many-arguments
+    # pylint: disable=too-few-public-methods
+    """
+    Helper class that will help a component manage entities of a platform.
+    """
+    def __init__(self, logger, hass, entity_id_format, group_obj,
+                 scan_interval, name):
+        self.logger = logger
+        self.hass = hass
+        self.entity_id_format = entity_id_format
+        self.entities = {}
+        self.scan_interval = scan_interval
+        self.is_polling = False
+
+        self.group = group_obj
+        self.name = name
+
+    def add_entities(self, new_entities):
+        """
+        Takes in a list of new entities. For each entity will see if it already
+        exists. If not, will add it, set it up and push the first state.
+        """
+        for entity in new_entities:
+            if entity is not None and entity not in self.entities.values():
+                entity.hass = self.hass
+
+                if getattr(entity, 'entity_id', None) is None:
+                    entity.entity_id = generate_entity_id(
+                        self.entity_id_format, entity.name,
+                        self.entities.keys())
+
+                self.entities[entity.entity_id] = entity
+
+                entity.update_ha_state()
+
+        if self.group is not None:
+            self.group.update_tracked_entity_ids(self.entities.keys())
```

Comment:
Calling `update_tracked_entity_ids` will overwrite any existing entities in that group. Updating the group should be handled by the EntityComponent which has access to all entities.